internal/remoteconfig/remoteconfig.go

```diff
@@ -0,0 +1,294 @@
+	PollRate time.Duration
+	// A list of remote config products this client is interested in
+	Products []string
+	// The tracer's runtime id
+	RuntimeID string
+	// The name of the user's application
+	ServiceName string
+	// The semantic version of the tracer
+	TracerVersion string
+	// The base TUF root metadata file
+	TUFRoot string
+	// The capabilities of the client
+	Capabilities []Capability
+}
+
+// A Client interacts with an Agent to update and track the state of remote
+// configuration
+type Client struct {
+	ClientConfig
+	client http.Client
+
+	clientID   string
+	repository *rc.Repository
+	stop       chan struct{}
+
+	callbacks map[string][]Callback
+
+	lastError error
+}
+
+// DefaultClientConfig generates a default remote config client configuration.
+// The default client configuration doesn't listen for any product.
+func DefaultClientConfig() ClientConfig {
+	return ClientConfig{
+		Env:           os.Getenv("DD_ENV"),
+		PollRate:      time.Second * 1,
+		RuntimeID:     globalconfig.RuntimeID(),
+		ServiceName:   globalconfig.ServiceName(),
+		TracerVersion: version.Tag,
+		TUFRoot:       os.Getenv("DD_RC_TUF_ROOT"),
+	}
+}
+
+// NewClient creates a new remoteconfig Client
+func NewClient(config ClientConfig) (*Client, error) {
+	repo, err := rc.NewRepository([]byte(config.TUFRoot))
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		ClientConfig: config,
+		clientID:     generateID(),
+		repository:   repo,
+		stop:         make(chan struct{}),
+		lastError:    nil,
+		callbacks:    map[string][]Callback{},
+	}, nil
+}
+
+// Start starts the client's update poll loop
+func (c *Client) Start() {
+	ticker := time.NewTicker(c.PollRate)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-c.stop:
+			return
+		case <-ticker.C:
+			c.updateState()
+		}
+	}
+}
+
+// Stop stops the client's update poll loop
+func (c *Client) Stop() {
+	close(c.stop)
+}
+
+func (c *Client) updateState() {
+	data, err := c.newUpdateRequest()
+	if err != nil {
+		c.lastError = err
+		return
+	}
+
+	url := fmt.Sprintf("http://%s/v0.7/config", c.AgentAddr)
+	req, err := http.NewRequest("GET", url, &data)
+	if err != nil {
+		c.lastError = err
+		return
+	}
+
+	resp, err := c.client.Do(req)
+	if err != nil {
+		c.lastError = err
+		return
+	}
+
+	var update clientGetConfigsResponse
+	err = json.NewDecoder(resp.Body).Decode(&update)
+	if err != nil {
+		c.lastError = err
+		return
+	}
+
+	err = c.applyUpdate(&update)
+	c.lastError = err
+}
+
+// RegisterCallback allows registering a callback that will be invoked when the client
+// receives a configuration update for the specified product.
+func (c *Client) RegisterCallback(f Callback, product string) {
+	c.callbacks[product] = append(c.callbacks[product], f)
+}
+
+func (c *Client) applyUpdate(pbUpdate *clientGetConfigsResponse) error {
+	fileMap := make(map[string][]byte, len(pbUpdate.TargetFiles))
+	for _, f := range pbUpdate.TargetFiles {
+		fileMap[f.Path] = f.Raw
+	}
```

Comment:
This is where I would filter them per product by simply changing the fileMap type to `map[string]map[string][]byte` so you can have product => file => content

Reply:
ba672a4053588